src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,430 @@
+import { assertIsValidKeyGetter, isCtrlKeyHeldDown } from './utils';
+import type {
+  CellKeyboardEvent,
+  CellKeyDownArgs,
+  GroupRow,
+  GroupRowHeightArgs,
+  Maybe,
+  Omit,
+  RenderRowProps,
+  RowHeightArgs,
+  RowsChangeData
+} from './types';
+import { renderToggleGroup, SELECT_COLUMN_KEY } from '.';
+import DataGrid from './DataGrid';
+import type { DataGridHandle, DataGridProps } from './DataGrid';
+import { useDefaultRenderers } from './DataGridDefaultRenderersProvider';
+import GroupedRow from './GroupRow';
+import { defaultRenderRow } from './Row';
+
+export interface TreeDataGridProps<R, SR = unknown, K extends Key = Key>
+  extends Omit<DataGridProps<R, SR, K>, 'role' | 'aria-rowcount' | 'rowHeight' | 'onFill'> {
+  rowHeight?: Maybe<number | ((args: GroupRowHeightArgs<R>) => number)>;
+  groupBy: readonly string[];
+  rowGrouper: (rows: readonly R[], columnKey: string) => Record<string, readonly R[]>;
+  expandedGroupIds: ReadonlySet<unknown>;
+  onExpandedGroupIdsChange: (expandedGroupIds: Set<unknown>) => void;
+}
+
+type GroupByDictionary<TRow> = Record<
+  string,
+  {
+    readonly childRows: readonly TRow[];
+    readonly childGroups: readonly TRow[] | Readonly<GroupByDictionary<TRow>>;
+    readonly startRowIndex: number;
+  }
+>;
+
+function TreeDataGrid<R, SR, K extends Key>(
+  {
+    columns: rawColumns,
+    rows: rawRows,
+    rowHeight: rawRowHeight,
+    rowKeyGetter: rawRowKeyGetter,
+    onCellKeyDown: rawOnCellKeyDown,
+    onRowsChange,
+    selectedRows: rawSelectedRows,
+    onSelectedRowsChange: rawOnSelectedRowsChange,
+    renderers,
+    groupBy: rawGroupBy,
+    rowGrouper,
+    expandedGroupIds,
+    onExpandedGroupIdsChange,
+    ...props
+  }: TreeDataGridProps<R, SR, K>,
+  ref: React.Ref<DataGridHandle>
+) {
+  const defaultRenderers = useDefaultRenderers<R, SR>();
+  const rawRenderRow = renderers?.renderRow ?? defaultRenderers?.renderRow ?? defaultRenderRow;
+  const headerAndTopSummaryRowsCount = 1 + (props.topSummaryRows?.length ?? 0);
+  const isRtl = props.direction === 'rtl';
+  const leftKey = isRtl ? 'ArrowRight' : 'ArrowLeft';
+  const rightKey = isRtl ? 'ArrowLeft' : 'ArrowRight';
+  const toggleGroupLatest = useLatestFunc(toggleGroup);
+
+  const { columns, groupBy } = useMemo(() => {
+    const columns = [...rawColumns].sort(({ key: aKey }, { key: bKey }) => {
+      // Sort select column first:
+      if (aKey === SELECT_COLUMN_KEY) return -1;
+      if (bKey === SELECT_COLUMN_KEY) return 1;
+
+      // Sort grouped columns second, following the groupBy order:
+      if (rawGroupBy.includes(aKey)) {
+        if (rawGroupBy.includes(bKey)) {
+          return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
+        }
+        return -1;
+      }
+      if (rawGroupBy.includes(bKey)) return 1;
+
+      // Sort other columns last:
+      return 0;
+    });
+
+    const groupBy: string[] = [];
+    for (const [index, column] of columns.entries()) {
```

Comment:
🤷‍♂️ 
```suggestion
    for (let index = 0; index < columns.length; index++ ) {
      const column = columns[index];
```

Reply:
Why is this preferred over `columns.entries()`

Reply:
Both are fine I guess, I'd expect the `for...let` loop to be more optimized though since it's an older pattern and doesn't create an iterator + an array on each iteration.
We can keep it as it is.